util/queue/chunkqueue.go

```diff
@@ -0,0 +1,401 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package queue
+
+import (
+	"fmt"
+	"sync"
+	"unsafe"
+)
+
+const (
+	// defaultSizePerChunk set the default size of each chunk be 1024 bytes (1kB)
+	defaultSizePerChunk = 1024
+	// minimumChunkLen is the minimum length of each chunk
+	minimumChunkLen = 16
+	// defaultPitchArrayLen is the default length of the chunk pointers array
+	defaultPitchArrayLen = 16
+)
+
+// ChunkQueue is a generic, efficient, iterable and GC-friendly queue.
+// Attention, it's not thread-safe.
+type ChunkQueue[T any] struct {
```

Comment:
ChunkQueue is based on the chunk. But I think it is called queue and it will be easier to understand.